File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	for len(bytes) > 0 {
		n, err := socket.Write(bytes)
		if err != nil {
			return err
		}
		bytes = bytes[n:]
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
    buff := make([]byte, size)
    received := 0
    for received < size {
        n, err := socket.Read(buff[received:])
        if err != nil {
            return nil, err
        }
        received += n
    }
    return buff, nil
}

File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size):
    bytes_recv = 0
    data = b""
    while bytes_recv < size:
        chunk = socket.recv(size - bytes_recv)
        if not chunk:
            raise ConnectionError("Socket connection closed before receiving all data")
        data += chunk
        bytes_recv += len(chunk)
    return data


def send_all(socket: socket.socket, bytes):
    while len(bytes) > 0:
        sent = socket.send(bytes)
        bytes = bytes[sent:]
    return socket.send(bytes)
